packages/lodestar/src/chain/genesis/genesis.ts

```diff
@@ -19,90 +25,112 @@ import {
   }
 
-  private processDepositEvents(): (source: AsyncIterable<Eth1EventsBlock>) => AsyncGenerator<[Deposit[], Eth1Block]> {
-    return (source) => {
-      const {depositTree, config} = this;
-      return (async function* () {
-        for await (const {events, block} of source) {
-          const newDeposits: Deposit[] = events.map((depositEvent) => {
-            depositTree.push(config.types.DepositData.hashTreeRoot(depositEvent));
-            return {
-              proof: depositTree.tree().getSingleProof(depositTree.gindexOfProperty(depositEvent.index)),
-              data: depositEvent,
-            };
-          });
-          yield [newDeposits, block] as [Deposit[], Eth1Block];
-        }
-      })();
-    };
+  /**
+   * First phase of waiting for genesis.
+   * Stream deposits events in batches as big as possible without querying block data
+   * @returns Block number at which there are enough active validators is state for genesis
+   */
+  private async waitForGenesisValidators(fromBlock: number): Promise<number> {
+    const controller = new AbortController();
+    const depositsStream = abortable(
+      getDepositsStream(fromBlock, this.eth1Provider, this.eth1Params),
+      controller.signal,
+      {returnOnAbort: true}
+    );
+
+    for await (const {depositEvents, blockNumber} of depositsStream) {
+      this.applyDeposits(depositEvents);
+      if (isValidGenesisValidators(this.config, this.state)) {
+        this.logger.info(`Found enough validators at eth1 block ${blockNumber}`);
+        controller.abort();
+        return blockNumber;
+      }
+    }
+
+    throw Error("depositsStream stopped without a valid genesis state");
   }
 
-  private async initialize(): Promise<void> {
-    const depositDatas = (await this.db.depositData.values()) || [];
-    const depositDataRoots = (await this.db.depositDataRoot.values()) || [];
-    depositDatas.map((event, index) => {
-      this.depositTree.push(depositDataRoots[index]);
-      return {
-        proof: this.depositTree.tree().getSingleProof(this.depositTree.gindexOfProperty(index)),
-        data: event,
-      };
-    });
+  private applyDeposits(depositEvents: IDepositEvent[]): void {
+    const newDeposits = depositEvents
+      .filter((depositEvent) => !this.depositCache.has(depositEvent.index))
+      .map((depositEvent) => {
+        this.depositCache.add(depositEvent.index);
+        this.depositTree.push(this.config.types.DepositData.hashTreeRoot(depositEvent));
+        return {
+          proof: this.depositTree.tree().getSingleProof(this.depositTree.gindexOfProperty(depositEvent.index)),
+          data: depositEvent,
+        };
+      });
+
+    applyDeposits(this.config, this.state, newDeposits, this.depositTree);
+
+    this.logger.verbose(`Found ${this.state.validators.length} validators to genesis`);
+
+    // TODO: If necessary persist deposits here to this.db.depositData, this.db.depositDataRoot
```

Comment:
IMO genesis builder can/should be stateless, not have access to `db`, just return data to be saved to db by its consumer.

We might not even need/want genesis builder to be a class:
idea:
```typescript
// in chain/chain.ts
try {
  const { state, depositTree, blockNumber } = await buildGenesis(
    this.eth1, this.logger, this.db.getGenesisWork()
  );
  ...
} catch (e) {
  if (e instanceof BuildGenesisError) {
    await this.db.saveGenesisWork(e.stuff);
  }
}

Reply:
Removed `db` from genesis builder modules